File: illusionflow.c
#include <stdint.h>

struct illusionflow64_s {
  uint64_t a;
  uint64_t b;
  uint64_t c;
  uint64_t d;
  uint64_t e;
};

struct illusionflow32_s {
  uint32_t a;
  uint32_t b;
  uint32_t c;
  uint32_t d;
  uint32_t e;
};

struct illusionflow16_s {
  uint16_t a;
  uint16_t b;
  uint16_t c;
  uint16_t d;
  uint16_t e;
};

struct illusionflow8_s {
  uint8_t a;
  uint8_t b;
  uint8_t c;
  uint8_t d;
  uint8_t e;
};

uint64_t illusionflow64(struct illusionflow64_s *s) {
  uint64_t mix = s->b + s->d;

  s->a++;

  if (s->a < 1) {
    s->c ^= s->b;
    s->e += s->d;
    s->b += s->e;
    s->d ^= s->c;
    return mix;
  }

  s->b = ((s->b << 35) | (s->b >> 29)) ^ s->c;
  s->c += 111111111111111111;
  s->d = (mix << 23) | (mix >> 41);
  return mix;
}

uint32_t illusionflow32(struct illusionflow32_s *s) {
  uint32_t mix = s->b + s->d;

  s->a++;

  if (s->a < 1) {
    s->c ^= s->b;
    s->e += s->d;
    s->b += s->e;
    s->d ^= s->c;
    return mix;
  }

  s->b = ((s->b << 17) | (s->b >> 15)) ^ s->c;
  s->c += 1111111111;
  s->d = (mix << 13) | (mix >> 19);
  return mix;
}

uint16_t illusionflow16(struct illusionflow16_s *s) {
  uint16_t mix = s->b + s->d;

  s->a++;

  if (s->a < 1) {
    s->c ^= s->b;
    s->e += s->d;
    s->b += s->e;
    s->d ^= s->c;
    return mix;
  }

  s->b = ((s->b << 9) | (s->b >> 7)) ^ s->c;
  s->c += 11111;
  s->d = (mix << 6) | (mix >> 10);
  return mix;
}

uint8_t illusionflow8(struct illusionflow8_s *s) {
  uint8_t mix = s->b + s->d;

  s->a++;

  if (s->a < 1) {
    s->c ^= s->b;
    s->e += s->d;
    s->b += s->e;
    s->d ^= s->c;
    return mix;
  }

  s->b = ((s->b << 3) | (s->b >> 5)) ^ s->c;
  s->c += 11;
  s->d = (mix << 6) | (mix >> 2);
  return mix;
}

File: illusionflow.cs
using System;

public class IllusionFlow64 {
  public ulong a;
  public ulong b;
  public ulong c;
  public ulong d;
  public ulong e;

  public ulong Next() {
    ulong mix = b + d;

    a++;

    if (a < 1) {
      c ^= b;
      e += d;
      b += e;
      d ^= c;
      return mix;
    }

    b = ((b << 35) | (b >> 29)) ^ c;
    c += 111111111111111111;
    d = (mix << 23) | (mix >> 41);
    return mix;
  }
}

public class IllusionFlow32 {
  public uint a;
  public uint b;
  public uint c;
  public uint d;
  public uint e;

  public uint Next() {
    uint mix = b + d;

    a++;

    if (a < 1) {
      c ^= b;
      e += d;
      b += e;
      d ^= c;
      return mix;
    }

    b = ((b << 17) | (b >> 15)) ^ c;
    c += 1111111111;
    d = (mix << 13) | (mix >> 19);
    return mix;
  }
}

public class IllusionFlow16 {
  public ushort a;
  public ushort b;
  public ushort c;
  public ushort d;
  public ushort e;

  public ushort Next() {
    ushort mix = (ushort) (b + d);

    a++;

    if (a < 1) {
      c ^= b;
      e += d;
      b += e;
      d ^= c;
      return mix;
    }

    b = (ushort) (((b << 9) | (b >> 7)) ^ c);
    c += 11111;
    d = (ushort) ((mix << 6) | (mix >> 10));
    return mix;
  }
}

public class IllusionFlow8 {
  public byte a;
  public byte b;
  public byte c;
  public byte d;
  public byte e;

  public byte Next() {
    byte mix = (byte) (b + d);

    a++;

    if (a < 1) {
      c ^= b;
      e += d;
      b += e;
      d ^= c;
      return mix;
    }

    b = (byte) (((b << 3) | (b >> 5)) ^ c);
    c += 11;
    d = (byte) ((mix << 6) | (mix >> 2));
    return mix;
  }
}
